pkg/migration/migrations/fix_privileges_migration_external_test.go

```diff
@@ -0,0 +1,218 @@
+2241083b1206736368656d61183d222b0a090a0561646d696e10020a080a04726f6f7410020a0c0a08746573747573657210201204726f6f7418012a00300140004a00
+`
+
+	// Ensure types can also be upgraded. Types in practice should not be corrupted
+	// and should always have an owner since they're created in 20.2 and onwards.
+	// We still want to upgrade the Version of the privilege descriptor.
+	// The following SQL statements were run in a 20.2.5 cluster.
+	/*
+		CREATE DATABASE db;
+		USE db;
+		CREATE TYPE t AS ENUM();
+
+		   The descriptors were then extracted using:
+
+		   	SELECT encode(descriptor, 'hex') AS descriptor
+		   		FROM system.descriptor
+		   	 WHERE id
+		   				 IN (
+		   							SELECT id
+		   								FROM system.namespace
+		   							 WHERE "parentID"
+		   										 = (
+		   													SELECT id
+		   														FROM system.namespace
+		   													 WHERE "parentID" = 0 AND name = 'db'
+		   											)
+		   									OR "parentID" = 0 AND name = 'db'
+		   					);
+	*/
+	const descriptorStringsToInjectSchemaTest = `
+122d0a026462103d1a1d0a090a0561646d696e10020a080a04726f6f7410021204726f6f7418012200280140004a00
+1a41083d101d1a0174203e2800403f480152006800722a0a090a0561646d696e10020a0b0a067075626c69631080040a080a04726f6f7410021204726f6f7418017a00
+1a6a083d101d1a025f74203f28013a26080f100018003000381850df8d065a14081810001800300050de8d0660007a0410df8d0660004000480152006800722a0a090a0561646d696e10020a0b0a067075626c69631080040a080a04726f6f7410021204726f6f7418017a00
+`
+
+	testCases := []string{
+		descriptorStringsToInjectZoneConfigFixTest,
+		descriptorStringsToInjectSchemaTest,
+		descriptorStringsToInjectUserDefinedTypeTest,
+	}
+	for _, descriptorStringsToInject := range testCases {
+		var descriptorsToInject []*descpb.Descriptor
+		for _, s := range strings.Split(strings.TrimSpace(descriptorStringsToInject), "\n") {
+			encoded, err := hex.DecodeString(s)
+			require.NoError(t, err)
+			var desc descpb.Descriptor
+			require.NoError(t, protoutil.Unmarshal(encoded, &desc))
+			descriptorsToInject = append(descriptorsToInject, &desc)
+		}
+
+		ctx := context.Background()
+		tc := testcluster.StartTestCluster(t, 1, base.TestClusterArgs{
+			ServerArgs: base.TestServerArgs{
+				Knobs: base.TestingKnobs{
+					Server: &server.TestingKnobs{
+						DisableAutomaticVersionUpgrade: 1,
+						BinaryVersionOverride:          clusterversion.ByKey(clusterversion.FixedPrivilegeDescriptors - 1),
+					},
+				},
+			},
+		})
+
+		db := tc.ServerConn(0)
+		kvDB := tc.Server(0).DB()
+		require.NoError(t, sqlutils.InjectDescriptors(ctx, db, descriptorsToInject, true))
+
+		_, err := tc.Conns[0].ExecContext(ctx, `SET CLUSTER SETTING version = $1`,
+			clusterversion.ByKey(clusterversion.FixedPrivilegeDescriptors).String())
+		require.NoError(t, err)
+
+		err = kvDB.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
+			descs, err := catalogkv.GetAllDescriptors(ctx, txn, keys.SystemSQLCodec)
```

Comment:
Seeing as `GetAllDescriptors` will validate those descriptors which will in turn validate the privilege descriptors, aren't some of the checks below redundant? I'm pretty sure the calls to `privilegeDesc.Validate` are. Speaking of which, shouldn't some checks be added to that method, perhaps version-gated behind `privilegeDesc.Version`?